File: services/busca/dominio/dominio_test.go
package dominio

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

// Os casos de busca vêm de contratos/exemplos/, CARREGADOS (não transcritos):
// app (Kotlin) e api (Java) carregam os mesmos arquivos, então um caso novo alcança
// as três implementações de uma vez. Havendo divergência, uma delas está errada.
// Ver ADR-0002 e contratos/exemplos/casos-de-busca.json.

// exemplos localiza contratos/exemplos/<nome> subindo a partir do diretório do teste,
// para funcionar tanto rodando `go test ./...` quanto a partir da raiz.
func exemplos(t *testing.T, nome string) []byte {
	t.Helper()
	dir, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	for {
		caminho := filepath.Join(dir, "contratos", "exemplos", nome)
		if _, err := os.Stat(caminho); err == nil {
			dados, err := os.ReadFile(caminho)
			if err != nil {
				t.Fatal(err)
			}
			return dados
		}
		pai := filepath.Dir(dir)
		if pai == dir {
			t.Fatalf("não encontrei contratos/exemplos/%s subindo de %s", nome, dir)
		}
		dir = pai
	}
}

func carregarAcervo(t *testing.T) []Obra {
	var arquivo struct {
		Obras []Obra `json:"obras"`
	}
	if err := json.Unmarshal(exemplos(t, "acervo.json"), &arquivo); err != nil {
		t.Fatal(err)
	}
	return arquivo.Obras
}

// filtroDeJSON converte a árvore JSON em Filtro. Vive no TESTE, não no domínio:
// converter JSON é trabalho de adaptador, e o domínio não conhece o formato.
func filtroDeJSON(t *testing.T, m map[string]any) Filtro {
	t.Helper()
	switch m["tipo"] {
	case "tem":
		return Tem{Dimensao: texto(m["dimensao"]), Valor: texto(m["valor"])}
	case "ou":
		return Ou{Opcoes: filtrosDeJSON(t, m["opcoes"])}
	case "e":
		return E{Exigencias: filtrosDeJSON(t, m["exigencias"])}
	case "exceto":
		return Exceto{Filtro: filtroDeJSON(t, m["filtro"].(map[string]any))}
	case "ate":
		return Ate{Ano: int(m["ano"].(float64))}
	default:
		t.Fatalf("tipo de filtro desconhecido: %v", m["tipo"])
		return nil
	}
}

func filtrosDeJSON(t *testing.T, v any) []Filtro {
	itens, _ := v.([]any)
	fs := make([]Filtro, len(itens))
	for i, it := range itens {
		fs[i] = filtroDeJSON(t, it.(map[string]any))
	}
	return fs
}

func texto(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}

func TestCasosCompartilhados(t *testing.T) {
	acervo := carregarAcervo(t)

	var arquivo struct {
		Casos []struct {
			Nome     string         `json:"nome"`
			Filtro   map[string]any `json:"filtro"`
			Esperado []string       `json:"esperado"`
		} `json:"casos"`
	}
	if err := json.Unmarshal(exemplos(t, "casos-de-busca.json"), &arquivo); err != nil {
		t.Fatal(err)
	}
	if len(arquivo.Casos) == 0 {
		t.Fatal("nenhum caso carregado de casos-de-busca.json")
	}

	for _, caso := range arquivo.Casos {
		t.Run(caso.Nome, func(t *testing.T) {
			obtido := make([]string, 0)
			for _, o := range Buscar(acervo, filtroDeJSON(t, caso.Filtro)) {
				obtido = append(obtido, o.ID)
			}
			if !reflect.DeepEqual(obtido, caso.Esperado) {
				t.Errorf("esperava %v, obteve %v", caso.Esperado, obtido)
			}
		})
	}
}

func TestDescreverAninhado(t *testing.T) {
	f := E{[]Filtro{
		Ou{[]Filtro{Tem{"ritmo", "ijexa"}, Tem{"ritmo", "baiao"}}},
		Exceto{Tem{"genero", "forro"}},
	}}
	esperado := "((ritmo=ijexa ou ritmo=baiao) e nao genero=forro)"
	if obtido := Descrever(f); obtido != esperado {
		t.Errorf("esperava %q, obteve %q", esperado, obtido)
	}
}

File: api-quarkus/src/test/java/br/ufrn/musi/DominioTest.java
package br.ufrn.musi;

import br.ufrn.musi.adaptadores.busca.Dtos;
import br.ufrn.musi.adaptadores.busca.Dtos.FiltroDto;
import br.ufrn.musi.adaptadores.busca.Dtos.ObraDto;
import br.ufrn.musi.dominio.Dominio.*;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestFactory;

import java.io.IOException;
import java.nio.file.Path;
import java.util.List;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import static br.ufrn.musi.dominio.Dominio.descrever;
import static br.ufrn.musi.dominio.Dominio.satisfaz;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.DynamicTest.dynamicTest;

/**
 * Os casos de busca são CARREGADOS de contratos/exemplos/, não transcritos. O
 * `FiltroTest.kt` de `shared/` e o `dominio_test.go` de `services/` carregam os MESMOS
 * arquivos, então um caso novo alcança as três implementações de uma vez. Se discordarem,
 * uma delas está errada — ver docs/decisoes/0001-stacks-e-estrutura.md.
 */
class DominioTest {

    private static final ObjectMapper JSON = new ObjectMapper();

    /** Acervo compartilhado, carregado uma vez. Usado tambem por BuscarObrasTest. */
    static final List<Obra> ACERVO = carregarAcervo();

    // Localiza contratos/exemplos/<nome> subindo a partir do diretorio do modulo.
    private static Path exemplos(String nome) {
        for (Path dir = Path.of("").toAbsolutePath(); dir != null; dir = dir.getParent()) {
            Path candidato = dir.resolve("contratos/exemplos/" + nome);
            if (candidato.toFile().exists()) {
                return candidato;
            }
        }
        throw new IllegalStateException("não encontrei contratos/exemplos/" + nome);
    }

    private static List<Obra> carregarAcervo() {
        try {
            JsonNode obras = JSON.readTree(exemplos("acervo.json").toFile()).get("obras");
            List<ObraDto> dtos = JSON.convertValue(obras,
                JSON.getTypeFactory().constructCollectionType(List.class, ObraDto.class));
            return dtos.stream().map(Dtos::paraDominio).toList();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @TestFactory
    Stream<DynamicTest> casosCompartilhados() throws IOException {
        JsonNode casos = JSON.readTree(exemplos("casos-de-busca.json").toFile()).get("casos");
        return StreamSupport.stream(casos.spliterator(), false).map(caso -> {
            String nome = caso.get("nome").asText();
            return dynamicTest(nome, () -> {
                Filtro filtro = Dtos.paraDominio(JSON.convertValue(caso.get("filtro"), FiltroDto.class));
                List<String> esperado = JSON.convertValue(caso.get("esperado"),
                    JSON.getTypeFactory().constructCollectionType(List.class, String.class));
                List<String> obtido = ACERVO.stream().filter(o -> satisfaz(o, filtro)).map(Obra::id).toList();
                assertEquals(esperado, obtido, nome);
            });
        });
    }

    @Test void descreverAninhado() {
        var f = new E(List.of(
            new Ou(List.of(new Tem("ritmo", "ijexa"), new Tem("ritmo", "baiao"))),
            new Exceto(new Tem("genero", "forro"))));
        assertEquals("((ritmo=ijexa ou ritmo=baiao) e não genero=forro)", descrever(f));
    }
}
